File: bjo14889.cpp
#include <iostream>
#include <cmath>
using namespace std;

int match[20][20];
bool visited[20];
int N;
int answer = 10000;

void solution(int cnt, int cur) {
	if (cnt == N / 2) {
		int result1 = 0;
		int result2 = 0;
		for (int i = 0; i < N; i++) {
			for (int j = i + 1; j < N; j++) {
				if (visited[i] && visited[j]) {
					result1 += (match[i][j] + match[j][i]);
				}
				else if (!visited[i] && !visited[j]) {
					result2 += (match[i][j] + match[j][i]);
				}
			}
		}
		answer = __min(answer, abs(result1 - result2));
	}
	for (int i = 0; i < N; i++) {
		if (visited[i] == false && i > cur) {
			visited[i] = true;
			solution(cnt + 1, i);
			visited[i] = false;
		}
	}
}

int main() {
	cin >> N;
	for (int i = 0; i < N; i++) {
		for (int j = 0; j < N; j++) {
			cin >> match[i][j];
		}
	}
	solution(0, -1);
	cout << answer << endl;
	system("pause");
}

File: bjo14889.py
from itertools import combinations

N=int(input())
match=[]
for _ in range(N):
    match.append(list(map(int, input().split())))
team_list=[i for i in range(N)]
answer=10000

def solution():
    global answer
    for comb in combinations(team_list, N//2):
        start=list(comb)
        link=list(set(team_list)-set(comb))

        s_comb=list(combinations(start, 2))
        l_comb=list(combinations(link, 2))

        result1=0
        for i,j in s_comb:
            result1+=(match[i][j]+match[j][i])
        result2=0
        for i,j in l_comb:
            result2+=(match[i][j]+match[j][i])

        if(answer>abs(result1-result2)):
            answer=abs(result1-result2)
solution()
print(answer)